rewrite.ml

```diff
@@ -0,0 +1,228 @@
+     let list = List.map (fun pc -> pc_map pc - 1) preds.(pc_above) in
+     Work (edit, List.sort Pervasives.compare list)
+   end
+  | _ as instr ->
+    if is_eliminating instr then Work (edit [|to_move|], [pc_above])
+    else if is_anihilating instr then Stop (edit [||])
+    else if is_blocking instr then Blocked
+    else Work (edit [|to_move; instr|], [pc_above])
+
+type pull_status =
+  | Pulled_to of Edit.result * pc
+  | Blocked
+
+let pull_instr instrs pc_branch : pull_status =
+  let succs = Analysis.successors instrs in
+  let instr_after_label pc =
+    assert (match[@warning "-4"] instrs.(pc) with
+        | Label _ -> true | _ -> false);
+    instr_at instrs (pc+1) in
+  let to_pull = List.map instr_after_label succs.(pc_branch) in
+  let sample_instr = List.hd to_pull in
+  let can_pull = List.for_all ((=) sample_instr) to_pull in
+  if not can_pull then Blocked
+  else begin
+    let insert = (pc_branch, 0, [| sample_instr |]) in
+    let delete label_pc = (label_pc + 1, 1, [||]) in
+    let deletes = List.map delete succs.(pc_branch) in
+    let (_instrs, pc_map) as edit = Edit.subst_many instrs (insert :: deletes) in
+    Pulled_to (edit, pc_map pc_branch - 1)
+  end
+
+type progress = Made_progress | Stuck
+
+let move_up_instr is_eliminating is_anihilating is_blocking instrs pc =
+  let push_instr = push_instr is_eliminating is_anihilating is_blocking in
+  let rec work_push instrs progress to_push to_pull =
+    match to_push with
+    | pc :: to_push ->
+      begin match push_instr instrs pc with
+        | Blocked ->
+          work_push instrs progress to_push to_pull
+        | Stop (instrs, pc_map) ->
+          let (!!) = List.map pc_map in
+          work_push instrs progress !!to_push !!to_pull
+        | Need_pull pc' ->
+          work_push instrs progress to_push (pc' :: to_pull)
+        | Work ((instrs, pc_map), pcs) ->
+          let (!!) = List.map pc_map in
+          work_push instrs Made_progress (List.rev_append pcs !!to_push) !!to_pull
+      end
+    | [] -> work_pull instrs progress to_pull []
+  and work_pull instrs progress to_pull blocked_branches =
+    match to_pull with
+    | pc :: to_pull ->
+      begin match pull_instr instrs pc with
+        | Pulled_to ((instrs, pc_map), pc') ->
+          let (!!) = List.map pc_map in
+          work_push instrs Made_progress [pc'] (List.rev_append !!blocked_branches !!to_pull)
+        | Blocked ->
+          work_pull instrs progress to_pull (pc :: blocked_branches)
+      end
+    | [] ->
+      match progress with
+      | Stuck ->
+        None
+      | Made_progress ->
+        Some instrs
+  in
+  work_push instrs Stuck pc []
+
+let move_up is_target is_eliminating is_anihilating is_blocking instrs =
+    let move_up_instr = move_up_instr is_eliminating is_anihilating is_blocking in
+
+    (* Loop around and move all drops until none moves anymore *)
+    let rec work status instrs pc =
```

Comment:
This function is a bit confusing, because `instrs` is reused. It's particularly bad in lines 171/172 where both cases return `instrs` but they're different. 177/178 is similar.